File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    int capacity;
    list<int> usageList;
    unordered_map<int, pair<int, list<int>::iterator>> cache;

    LRUCache(int capacity) : capacity(capacity) {}

    int get(int key) {
        if (cache.find(key) == cache.end()) {
            return -1;
        }
        usageList.erase(cache[key].second);
        usageList.push_front(key);
        cache[key].second = usageList.begin();
        return cache[key].first;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            usageList.erase(cache[key].second);
            usageList.push_front(key);
            cache[key] = {value, usageList.begin()};
            return;
        }
        if (cache.size() == capacity) {
            int lruKey = usageList.back();
            usageList.pop_back();
            cache.erase(lruKey);
        }
        usageList.push_front(key);
        cache[key] = {value, usageList.begin()};
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: LeetCode/0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._remove(node)
        self._insert(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self._remove(self.cache[key])
        if len(self.cache) == self.capacity:
            self._remove(self.tail.prev)
        new_node = Node(key, value)
        self._insert(new_node)

    def _remove(self, node: Node):
        del self.cache[node.key]
        node.prev.next = node.next
        node.next.prev = node.prev

    def _insert(self, node: Node):
        self.cache[node.key] = node
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)